source/common/upstream/original_dst_cluster.cc

```diff
@@ -88,11 +98,30 @@ HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerCont
       }
     }
   }
-
+  // TODO(ramaraochavali): add a stat and move this log line to debug.
   ENVOY_LOG(warn, "original_dst_load_balancer: No downstream connection or no original_dst.");
   return nullptr;
 }
 
+Network::Address::InstanceConstSharedPtr
+OriginalDstCluster::LoadBalancer::requestOverrideHost(LoadBalancerContext* context) {
+  Network::Address::InstanceConstSharedPtr request_host;
+  const Http::HeaderMap* downstream_headers = context->downstreamHeaders();
+  if (downstream_headers &&
+      downstream_headers->get(Http::Headers::get().EnvoyOriginalDstHost) != nullptr) {
+    const std::string& request_override_host =
+        downstream_headers->get(Http::Headers::get().EnvoyOriginalDstHost)->value().c_str();
+    try {
+      request_host = Network::Utility::parseInternetAddressAndPort(request_override_host, false);
+      ENVOY_LOG(debug, "Using request override host {}.", request_override_host);
+    } catch (const Envoy::EnvoyException& e) {
+      ENVOY_LOG(debug, "original_dst_load_balancer: invalid override header value. {}", e.what());
+      parent_.lock()->info_->stats().original_dst_host_invalid_.inc();
```

Comment:
I don't believe it's guaranteed that the `lock()` will actually return a valid pointer. What you should do instead is store a shared pointer to the cluster info inside this and use that.

Reply:
Ok. Looks like sharedptr is already available. I will use it.